Add Prometheus gRPC metrics for hubble and hubble-relay

chancez opened the pull request:
I wasn't sure if I should split this up into multiple PRs (one for hubble, one for relay, and maybe more for just adding interceptors vs the metrics middleware), but I didn't because then I would have to deal with adding the same dependency in each and deal with conflicts later, and the code is quite similar for the two subsystems. Let me know if you would prefer this is split up.

This PR:

- Does a bit of refactoring for the hubble server to follow common Go conventions, and to support gRPC interceptors.
- Adds support for gRPC interceptors on hubble relay
- Adds a prometheus metrics endpoint to hubble relay
- Adds prometheus gRPC middleware to both hubble and hubble relay

These metrics can help determine if errors are occurring with the hubble APIs, and diagnose potential performance impacts of making queries against hubble.

Additionally, by supporting configuring gRPC interceptors, we can further enhance our gRPC servers with additional middleware (eg: logging, tracing, auth, etc) in the future.

```release-note
Add Prometheus gRPC metrics for hubble and hubble-relay
```

chancez committed:
hubble: Make (*Server).Serve() blocking

In Go, Serve() blocks, and it's the callers responsibiilty to decide if
it should run in a Go routine.

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

chancez committed:
hubble: Support configuring gRPC interceptors

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

chancez committed:
relay: Move gRPC Server initialization into (*Server).New()

Too much initialization was occurring in the Serve() function which
should generally only deal with listening on the socket/port and
starting the server.

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

chancez committed:
relay: Support configuring gRPC interceptors

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

chancez committed:
hubble: Add Prometheus metrics for gRPC APIs

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

chancez committed:
relay: Add Prometheus metrics endpoint

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

chancez committed:
relay: Add Prometheus metrics for gRPC APIs

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

gandro commented on pkg/hubble/observer/local_observer.go:
```diff
@@ -105,8 +105,7 @@ func NewLocalServer(
 func (s *LocalObserverServer) Start() {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+	ctx := context.TODO()
```
I agree that this currently unused, but the intent here is to allow the hooks (e.g. `OnDecodedEvent`) to store state in the `ctx` and potentially even spawn go routines that we want to stop when the the server is stopped. 

The server can be stopped by calling `close(s.GetEventsChannel())` (which of course requires coordination with the sender). This is currently only done in unit tests. 

I'm fine with removing it, but maybe we should add a comment here in case we ever implement the shutdown logic?

chancez replied:
Ah interesting. So if a hook spawns a go routine, then when we return from `Start()` the deferred  `cancel()` runs, and if the go routine is watching the `ctx.Done()` channel, it should receive the cancellation. I can revert this commit and add a comment to mention this. That's something I hadn't even considered.

gandro replied:
Thanks for documenting this!

gandro commented on daemon/cmd/hubble.go:
```diff
@@ -191,10 +191,11 @@ func (d *Daemon) launchHubble() {
 	}
 	logger.WithField("address", sockPath).Info("Starting local Hubble server")
-	if err := localSrv.Serve(); err != nil {
-		logger.WithError(err).Error("Failed to start local Hubble server")
-		return
```
So here the old code `return`s if there is an error, ensuring no subsequent code is executed. 

The new code doesn't, but it moves a lot of code into the initialization, so this seems fine for the most part. The only remaining case where `Serve()` returns an error seems to be `errNoListener`, which is already checked in initialization here:
https://github.com/cilium/cilium/blob/3e2e95fe6a28af94946d7b944d9cd61044296d63/pkg/hubble/server/server.go#L43-L45

I don't see any code path which can modify the Listener after `NewServer` new server has been called, so I believe the check in `Serve()` is bogus and could be removed as well.

This means this log line here is essentially is never executed (since the new implementation of `Serve()` never returns any error), maybe we should fully remove the return value on `Serve()` and just have a `go localSrv.Server()` here?

chancez replied:
There was no code relying on localSrv after this, so my thinking was by not returning, we allow more of the hubble server to start, minimizing how degraded the functionality is when it can't start 1 of the servers. 

Regarding the error, I actually kind of think Serve should probably just return the error and let the caller handle logging it. 

Additionally, I don't agree with the idea of having it call Serve() in s go routine. This is what it used to do, but Go conventions are that blocking calls should be launched in Go routines by the caller. This is actually why I moved much of the logic from Serve() into NewServer().

gandro replied:
> Additionally, I don't agree with the idea of having it call Serve() in s go routine. This is what it used to do, but Go conventions are that blocking calls should be launched in Go routines by the caller. This is actually why I moved much of the logic from Serve() into NewServer().

Not sure what you mean by this. I agree with your appraoch which moves the go routine out of `Serve()` and to the caller, I think that's an improvement. My concern is solely about the inconsistent error handling and unused log lines.

chancez replied:
Ok, then we're on the same page. I'm going to refactor this again slightly to just return the error and then this code won't be unused and we'll have the logging done by the caller.

chancez replied:
Error is now propagated up

gandro commented on pkg/hubble/server/server.go:
```diff
@@ -90,17 +94,12 @@ func (s *Server) initGRPCServer() error {
-	}
 	if s.opts.Listener == nil {
 		return errNoListener
 	}
```
As mentioned above, let's remove this too.

gandro commented on pkg/hubble/server/server.go:
```diff
@@ -90,17 +94,12 @@ func (s *Server) initGRPCServer() error {
-	}(s.opts.Listener)
+	if err := s.srv.Serve(s.opts.Listener); err != nil {
+		s.log.WithError(err).WithField("address", s.opts.Listener.Addr().String()).Error("Failed to start gRPC server")
+	}
```
As mentioned above, with the `initGRPCServer` check moved to the constructor, we should either always propagate the error to the caller (i.e. remove this log line, but keep the current `func Serve() error` return value) or never (i.e. keep the log line, but remove the error return value from `Serve()`).

I'm fine either way.

chancez replied:
Addressed this by propagating the error.

gandro commented on pkg/hubble/server/server.go:
```diff
@@ -55,18 +55,20 @@ func NewServer(log logrus.FieldLogger, options ...serveroption.Option) (*Server,
 
 func (s *Server) newGRPCServer() (*grpc.Server, error) {
-	switch {
-	case s.opts.Insecure:
-		return grpc.NewServer(), nil
```
The commit message doesn't mention anything about this removal. I assume this is fine to be removed because we already check that `ServerTLSConfig` is not nil in the constructor?

chancez replied:
Yeah, this was always redundant from what I could tell. With the refactoring moving this into the constructor, it handles this and you can't modify listener after creating the Server, so it isn't needed.

gandro commented on pkg/hubble/relay/server/server.go:
```diff
@@ -99,39 +124,13 @@ func New(options ...Option) (*Server, error) {
-		s.server = grpc.NewServer(grpc.Creds(creds))
-	default:
-		return ErrNoServerTLSConfig
-	}
```
This error message is removed by the refactor. Please add it back in to `New()`. I believe this error is thrown intentionally, such that forgetting the TLS config cannot be done without explicitly requesting an insecure server.

chancez replied:
This should be handled here I believe? https://github.com/cilium/cilium/pull/20376/files#diff-c559920792798792688868000d2e056dc5ca3eed9c7a973ae73e8aea39e0bd7dL54-L59

gandro replied:
The link somehow didn't work for me, but judging from the URL you likely were referring to this:
https://github.com/cilium/cilium/blob/5b51ba8784ee30d9de85212fff8283ac93a4da81/pkg/hubble/relay/server/server.go#L56-L58

Yup, agreed that this should be handled. Thanks!

chancez replied:
Such is the pain of GitHub reviews and force pushing sometimes.

gandro commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -898,6 +898,25 @@ hubble:
     # port 80 if not.
     # servicePort: 80
 
+    # -- Enable prometheus metrics for cilium-operator on the configured port at
```
This doc comment mentions `cilium-operator`, I believe it should be `hubble-relay`

gandro reviewed (changes requested):
Awesome work. I'm not familiar with interceptors or the Prometheus gRPC metrics package, so I mostly focused around the existing Hubble logic. A few minor things that I think need to be addressed.

chancez committed:
hubble: Document purpose of cancellation context

The cancel() looks potentially unused in the local observer code here,
so added a comment indicating how it's used and why the cancellation is
there at all.

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

chancez committed:
relay: Support adding to default server options

This is based off the same approach we take within
observer.NewLocalServer, which allows other packages to extend the
default options of Hubble relay the same way packages can extend the
options of the Hubble observer server.

Signed-off-by: Chance Zibolski <chance.zibolski@gmail.com>

gandro reviewed (approved):
Awesome, this looks great. Thanks a lot!

sayboras commented on install/kubernetes/cilium/templates/hubble-relay/servicemonitor.yaml:
```diff
@@ -0,0 +1,26 @@
+  endpoints:
+  - port: metrics
+    interval: {{ .Values.hubble.relay.prometheus.serviceMonitor.interval | quote }}
+    path: /metrics
```
should we add  `honorLabels: true` to be aligned with other ServiceMonitor ?

chancez replied:
We shouldn't generally need that in most cases, it somewhat goes against best practices. I think it's only needed for daemonsets so some node specific labels are honored, but even then I'd argue we should just be relabeling them instead. 

We definitely shouldn't need it for relay IMO.

sayboras commented on install/kubernetes/cilium/templates/hubble-relay/metrics_service.yaml:
```diff
@@ -0,0 +1,20 @@
+{{- if and .Values.hubble.enabled .Values.hubble.relay.enabled .Values.hubble.relay.prometheus.enabled }}
```
nit: file name could be metrics-service.yaml for consistency.

chancez replied:
Good catch. Updated.

sayboras reviewed (approved):
Very comprehensive PR :100: , I have only two minor comments as per below.

Sample metrics list for other reviewer

<details>

```
# HELP grpc_server_handled_total Total number of RPCs completed on the server, regardless of success or failure.
# TYPE grpc_server_handled_total counter
grpc_server_handled_total{grpc_code="Aborted",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Aborted",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="AlreadyExists",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Canceled",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DataLoss",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="DeadlineExceeded",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="FailedPrecondition",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Internal",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="InvalidArgument",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="NotFound",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OK",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="OutOfRange",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="PermissionDenied",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="ResourceExhausted",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unauthenticated",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unavailable",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unimplemented",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_handled_total{grpc_code="Unknown",grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
# HELP grpc_server_msg_received_total Total number of RPC stream messages received on the server.
# TYPE grpc_server_msg_received_total counter
grpc_server_msg_received_total{grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_msg_received_total{grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_msg_received_total{grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_msg_received_total{grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_msg_received_total{grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_msg_received_total{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_msg_received_total{grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_msg_received_total{grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
# HELP grpc_server_msg_sent_total Total number of gRPC stream messages sent by the server.
# TYPE grpc_server_msg_sent_total counter
grpc_server_msg_sent_total{grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_msg_sent_total{grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_msg_sent_total{grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_msg_sent_total{grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_msg_sent_total{grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_msg_sent_total{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_msg_sent_total{grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_msg_sent_total{grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
# HELP grpc_server_started_total Total number of RPCs started on the server.
# TYPE grpc_server_started_total counter
grpc_server_started_total{grpc_method="Check",grpc_service="grpc.health.v1.Health",grpc_type="unary"} 0
grpc_server_started_total{grpc_method="GetAgentEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_started_total{grpc_method="GetDebugEvents",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_started_total{grpc_method="GetFlows",grpc_service="observer.Observer",grpc_type="server_stream"} 0
grpc_server_started_total{grpc_method="GetNodes",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_started_total{grpc_method="ServerReflectionInfo",grpc_service="grpc.reflection.v1alpha.ServerReflection",grpc_type="bidi_stream"} 0
grpc_server_started_total{grpc_method="ServerStatus",grpc_service="observer.Observer",grpc_type="unary"} 0
grpc_server_started_total{grpc_method="Watch",grpc_service="grpc.health.v1.Health",grpc_type="server_stream"} 0
```

</details>

rolinh commented on pkg/hubble/relay/server/option.go:
```diff
@@ -171,3 +189,28 @@ func WithLocalClusterName(clusterName string) Option {
+}
+
+// WithGRPCStreamInterceptor configures the server with the given gRPC server stream interceptor
+func WithGRPCStreamInterceptor(interceptor grpc.StreamServerInterceptor) Option {
```
Should we make this function variadic to allow passing several interceptors?

chancez replied:
That's a good idea.

chancez replied:
Updated

rolinh commented on pkg/hubble/relay/server/option.go:
```diff
@@ -171,3 +189,28 @@ func WithLocalClusterName(clusterName string) Option {
+}
+
+// WithGRPCUnaryInterceptor configures the server with the given gRPC server stream interceptor
+func WithGRPCUnaryInterceptor(interceptor grpc.UnaryServerInterceptor) Option {
```
ditto

chancez replied:
Updated

rolinh commented on pkg/hubble/server/serveroption/option.go:
```diff
@@ -134,3 +139,28 @@ func WithRecorderService(svc recorderpb.RecorderServer) Option {
+}
+
+// WithGRPCStreamInterceptor configures the server with the given gRPC server stream interceptor
+func WithGRPCStreamInterceptor(interceptor grpc.StreamServerInterceptor) Option {
```
ditto

chancez replied:
Updated

rolinh commented on pkg/hubble/server/serveroption/option.go:
```diff
@@ -134,3 +139,28 @@ func WithRecorderService(svc recorderpb.RecorderServer) Option {
+}
+
+// WithGRPCUnaryInterceptor configures the server with the given gRPC server stream interceptor
+func WithGRPCUnaryInterceptor(interceptor grpc.UnaryServerInterceptor) Option {
```
ditto

chancez replied:
Updated

rolinh reviewed (approved):
Excellent :rocket: 
I left a non-blocking suggestion.
